solidity/test/token-distribution/TokenDistributorTest.js

```diff
@@ -0,0 +1,651 @@
+          0,
+          unclaimedUnlockDuration,
+          {
+            from: owner,
+          }
+        ),
+        "amount has to be greater than zero"
+      )
+    })
+
+    it("reverts on token transfer failure", async function () {
+      await expectRevert(
+        tokenDistributor.allocate(
+          testData.merkle.merkleRoot,
+          testData.merkle.tokenTotal + 1,
+          unclaimedUnlockDuration,
+          {
+            from: owner,
+          }
+        ),
+        "SafeERC20: low-level call failed"
+      )
+    })
+
+    it("reverts when called by non-owner", async function () {
+      await expectRevert(
+        tokenDistributor.allocate(
+          testData.merkle.merkleRoot,
+          testData.merkle.tokenTotal,
+          unclaimedUnlockDuration,
+          {
+            from: thirdParty,
+          }
+        ),
+        "Ownable: caller is not the owner"
+      )
+    })
+  })
+
+  describe("claim", async () => {
+    before(async () => {
+      await tokenDistributor.allocate(
+        testData.merkle.merkleRoot,
+        testData.merkle.tokenTotal,
+        unclaimedUnlockDuration,
+        { from: owner }
+      )
+    })
+
+    after(freshDeployment)
+
+    it("transfers tokens", async function () {
+      const recipientInitialBalance = toBN(await testToken.balanceOf(recipient))
+      const destinationInitialBalance = toBN(
+        await testToken.balanceOf(claimDestination)
+      )
+
+      await tokenDistributor.claim(
+        recipient,
+        claimDestination,
+        testData.signature.v,
+        testData.signature.r,
+        testData.signature.s,
+        testData.merkle.claims[recipient].index,
+        testData.merkle.claims[recipient].amount,
+        testData.merkle.claims[recipient].proof
+      )
+
+      expect(
+        await testToken.balanceOf(recipient),
+        "invalid recipient address balance"
+      ).to.eq.BN(toBN(recipientInitialBalance))
+
+      expect(
+        await testToken.balanceOf(claimDestination),
+        "invalid destination address balance"
+      ).to.eq.BN(
+        destinationInitialBalance.add(
+          toBN(testData.merkle.claims[recipient].amount)
+        )
+      )
+    })
+
+    it("emits event", async function () {
+      expectEvent(
+        await tokenDistributor.claim(
+          recipient,
+          claimDestination,
+          testData.signature.v,
+          testData.signature.r,
+          testData.signature.s,
+          testData.merkle.claims[recipient].index,
+          testData.merkle.claims[recipient].amount,
+          testData.merkle.claims[recipient].proof
+        ),
+        "TokensClaimed",
+        {
+          index: testData.merkle.claims[recipient].index.toString(),
+          recipient: recipient,
+          destination: claimDestination,
+          amount: toBN(testData.merkle.claims[recipient].amount),
+        }
+      )
+    })
+
+    describe("when verifying destination signature", async function () {
```

Comment:
Another way of testing an internal `recoverSignerAddress()` would be to create a contract stub with a public function that calls an internal one. It can make it more focused. But we can leave it as is, up to you.

Reply:
Yep, this could be an alternative way to test it, we use it quite often for internal functions tests isolation. Here I decided to not add another stub contract as the current way of testing doesn't hurt that much in my opinion and we can easily test `recoverSignerAddress` results with the exposed function.